numpy/core/src/common/simd/sse/arithmetic.h

```diff
@@ -91,6 +91,125 @@ NPY_FINLINE __m128i npyv_mul_u8(__m128i a, __m128i b)
+    }
+    else {
+        m = 1;                                                  // for d1 = 1
+        sh = 0;
+        if (d == 0) {
+            m /= d;  // provoke error here if d = 0
+        }
+        if (d == 0x8000) { // fix overflow for this special case
+            m = 0x8001;
+            sh = 14;
+        }
+    }
+    npyv_divisor_t divisor;
+    divisor.val[0] = _mm_set1_epi16((npy_int16)m);           // broadcast multiplier
+    divisor.val[1] = _mm_setr_epi32(sh, 0, 0, 0);            // shift count
+    divisor.val[2] = _mm_set1_epi32(d < 0 ? -1 : 0);         // sign of divisor
+    return divisor;
+}
+
+NPY_FINLINE npyv_divisor_t npyv_divisor_u16(npy_uint16 d)
+{
+    npy_uint16 L, L2, sh1, sh2, m;
+    switch (d) {
+    case 0:
+        m = sh1 = sh2 = 1u / d;                        // provoke error for d = 0
+        break;
+    case 1:
+        m = 1; sh1 = sh2 = 0;                          // parameters for d = 1
+        break;
+    case 2:
+        m = 1; sh1 = 1; sh2 = 0;                       // parameters for d = 2
+        break;
+    default:                                           // general case for d > 2
+        L = npy_bit_scan_reverse_u32(d - 1u) + 1u;     // ceil(log2(d))
+        L2 = (1 << L);                                 // 2^L, overflow to 0 if L = 16
+        m = 1u + (((L2 - d) << 16) / d);               // multiplier
+        sh1 = 1;  sh2 = L - 1u;                        // shift counts
+    }
+    npyv_divisor_t divisor;
+    divisor.val[0] = _mm_set1_epi16((npy_int16)m);
+    divisor.val[1] = _mm_setr_epi32((npy_int32)sh1, 0, 0, 0);
+    divisor.val[2] = _mm_setr_epi32((npy_int32)sh2, 0, 0, 0);
+    return divisor;
+}
+
+NPY_FINLINE npyv_divisor_t npyv_divisor_u32(npy_uint32 d)
+{
+    npy_uint32 L, L2, sh1, sh2, m;
+    switch (d) {
+    case 0:
+        m = sh1 = sh2 = 1 / d;                         // provoke error for d = 0
+        break;
+    case 1:
+        m = 1; sh1 = sh2 = 0;                          // parameters for d = 1
+        break;
+    case 2:
+        m = 1; sh1 = 1; sh2 = 0;                       // parameters for d = 2
+        break;
+    default:                                           // general case for d > 2
+        L = npy_bit_scan_reverse_u32(d - 1) + 1;       // ceil(log2(d))
+        L2 = (L < 32 ? 1 << L : 0);                    // 2^L, overflow to 0 if L = 32
+        m = 1 + (((npy_uint64)(L2 - d) << 32) / d);    // multiplier
+        sh1 = 1;  sh2 = L - 1;                         // shift counts
+    }
+    npyv_divisor_t divisor;
+    divisor.val[0] = _mm_set1_epi32((npy_int32)m);
+    divisor.val[1] = _mm_setr_epi32((npy_int32)sh1, 0, 0, 0);
+    divisor.val[2] = _mm_setr_epi32((npy_int32)sh2, 0, 0, 0);
+    return divisor;
+}
+
+NPY_FINLINE npyv_s16 npyv_divc_s16(npyv_s16 a, const npyv_divisor_t divisor)
+{
+    __m128i t1 = _mm_mulhi_epi16(a, divisor.val[0]);   // multiply high signed words
```

Comment:
If I replace `__m*` with proper `npyv`, can I move it to simd_utils.h or a similar file? That way we do not rewrite for all AVX, AVX512, etc?

Reply:
wouldn't be flexible plus `npyv` doesn't cover operations like mul expand.